test/cpp/c10d/ProcessGroupNCCLErrorsTest.cpp

```diff
@@ -257,3 +345,112 @@ TEST_F(ProcessGroupNCCLErrorsTest, testNCCLErrorsNonBlocking) {
 
   // Communicators might be aborted here, further operations would fail.
 }
+
+TEST_F(ProcessGroupNCCLErrorsTest, testNCCLErrorsNoHeartbeat) {
+  if (skipTest()) {
+    return;
+  }
+
+  int heartBeatIntervalInSec = 2;
+  std::string timeInterval = std::to_string(heartBeatIntervalInSec);
+  ASSERT_TRUE(setenv(c10d::NCCL_BLOCKING_WAIT, "1", 1) == 0);
+  ASSERT_TRUE(
+      setenv(c10d::TORCH_NCCL_HEARTBEAT_TIMEOUT_SEC, timeInterval.c_str(), 1) ==
+      0);
+  ASSERT_TRUE(setenv(c10d::TORCH_NCCL_ENABLE_MONITORING, "1", 1) == 0);
+  auto options = c10d::ProcessGroupNCCL::Options::create();
+  // Set a long watchdog timeout, so that we have enough time to lock the
+  // watchdog and let the heartbeat monitor thread to kick in.
+  options->timeout = std::chrono::milliseconds(30000);
+  ProcessGroupNCCLNoHeartbeatCaught pg(store_, 0, 1, options);
+
+  // Normal collective case.
+  auto work = pg.allreduce(tensors_);
+  work->wait();
+  EXPECT_TRUE(work->isSuccess());
+
+  work = pg.allreduce(tensors_);
+  {
+    // Now run all reduce with errors.
+    std::lock_guard<std::mutex> lock(pg.getWatchdogMutex());
+    LOG(INFO) << "Lock watchdog thread.";
+    // Wait long enough before monitor thread throws exceptions.
+    std::this_thread::sleep_for(
+        std::chrono::seconds(heartBeatIntervalInSec * 3));
+    // Check the monitoring thread launched and exception thrown.
+    EXPECT_TRUE(pg.getErrorCaughtFlag());
+  }
+  work->wait();
+  EXPECT_TRUE(work->isSuccess());
+}
+
+class ProcessGroupNCCLWatchdogTimeoutTest : public ProcessGroupNCCLErrorsTest {
+ protected:
+  void watchdogTimeoutTestStepup() {
```

Comment:
nit: is this supposed to be named 'Setup' instead of 'Stepup'?

Reply:
opps, type.. My bad